readthedocs/projects/tasks/builds.py

```diff
@@ -550,11 +588,19 @@ def on_retry(self, exc, task_id, args, kwargs, einfo):
             self.update_build(state=BUILD_STATE_TRIGGERED)
 
     def after_return(self, status, retval, task_id, args, kwargs, einfo):
+        """
+        Celery handler to be executed after a task runs.
+
+        .. note::
+
+           This handler is called even if the task has failed,
+           so some attributes from the `self.data` object may not be defined.
+        """
         # Update build object
         self.data.build['length'] = (timezone.now() - self.data.start_time).seconds
 
         build_state = None
-        if self.data.build["state"] not in BUILD_FINAL_STATES:
+        if self.data.build.get("state") not in BUILD_FINAL_STATES:
```

Comment:
this is the only attribute from `data.build` that may not be defined, all others are defined in `on_failure` or inside this same method.